mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -0,0 +1,422 @@
+use std::time::Instant;
+use std::{ops::Range, sync::Arc};
+
+use bytes::Bytes;
+use futures::task::{Spawn, SpawnExt};
+use futures::{pin_mut, StreamExt};
+use mountpoint_s3_client::{types::ETag, ObjectClient};
+use tracing::{debug_span, error, trace, warn, Instrument};
+
+use crate::checksums::ChecksummedBytes;
+use crate::data_cache::{BlockIndex, CacheKey, DataCache};
+use crate::prefetch::part::Part;
+use crate::prefetch::part_queue::{unbounded_part_queue, PartQueueProducer};
+use crate::prefetch::part_stream::{ObjectPartStream, RequestRange};
+use crate::prefetch::task::RequestTask;
+use crate::prefetch::PrefetchReadError;
+
+/// [ObjectPartStream] implementation which maintains a [DataCache] for the object data
+/// retrieved by an [ObjectClient].
+#[derive(Debug)]
+pub struct CachingPartStream<Cache, Runtime> {
+    cache: Arc<Cache>,
+    runtime: Runtime,
+}
+
+impl<Cache, Runtime> CachingPartStream<Cache, Runtime> {
+    pub fn new(runtime: Runtime, cache: Cache) -> Self {
+        Self {
+            cache: Arc::new(cache),
+            runtime,
+        }
+    }
+}
+
+impl<Cache, Runtime> ObjectPartStream for CachingPartStream<Cache, Runtime>
+where
+    Cache: DataCache + Send + Sync + 'static,
+    Runtime: Spawn,
+{
+    fn spawn_get_object_request<Client>(
+        &self,
+        client: &Client,
+        bucket: &str,
+        key: &str,
+        if_match: ETag,
+        range: RequestRange,
+        _preferred_part_size: usize,
+    ) -> RequestTask<<Client as ObjectClient>::ClientError>
+    where
+        Client: ObjectClient + Clone + Send + Sync + 'static,
+    {
+        let range = get_aligned_request_range(range, self.cache.block_size());
+
+        let start = range.start();
+        let size = range.len();
+
+        let (part_queue, part_queue_producer) = unbounded_part_queue();
+        trace!(?range, "spawning request");
+
+        let request_task = {
+            let request = CachingRequest::new(
+                client.clone(),
+                self.cache.clone(),
+                bucket.to_owned(),
+                key.to_owned(),
+                if_match,
+                part_queue_producer,
+            );
+            let span = debug_span!("prefetch", ?range);
+            async move {
+                request.get_from_cache(range).await;
+            }
+            .instrument(span)
+        };
+
+        let task_handle = self.runtime.spawn_with_handle(request_task).unwrap();
+
+        RequestTask::from_handle(task_handle, size, start, part_queue)
+    }
+}
+
+fn get_aligned_request_range(range: RequestRange, block_size: u64) -> RequestRange {
```

Comment:
Sort of frustrating that this tricky part alignment logic is almost, but not quite, copied and pasted from the non-caching version. Actually, isn't it exactly the same except for the last case? Maybe we should write a common version that returns `Option<RequestRange>` and then have different fallback logic in the two cases?